### PR TITLE
feat(entity): add support for pagination in query entities #388

### PR DESCRIPTION
**Pull request by HoucemKacem**

### CODE REVIEW

File: entity-service/src/main/kotlin/com/egm/stellio/entity/web/EntityHandler.kt [hunk 1]
```diff
@@ -94,6 +97,10 @@ class EntityHandler(
         val contextLink = getContextFromLinkHeaderOrDefault(httpHeaders)
         val mediaType = getApplicableMediaType(httpHeaders)
 
+        if (limit <= 0 || page <= 0)
+            return ResponseEntity.status(HttpStatus.BAD_REQUEST).contentType(MediaType.APPLICATION_JSON)
+                .body(BadRequestDataResponse("Page number and Limit must be greater than zero"))
```
Comment (bobeal) [inline entity-service/src/main/kotlin/com/egm/stellio/entity/web/EntityHandler.kt:102]: greater or strictly greater? what if I set limit to 0? page starts at 0 or at 1?

File: entity-service/src/main/kotlin/com/egm/stellio/entity/web/EntityHandler.kt [hunk 2]
```diff
@@ -118,9 +125,11 @@ class EntityHandler(
                 userId
             ).toListOfString()
 
+        val entitiesChunk = entitiesUserCanRead.chunked(limit).getOrNull(page - 1) ?: emptyList()
```
Comment (bobeal) [inline entity-service/src/main/kotlin/com/egm/stellio/entity/web/EntityHandler.kt:128]: this has a major performance problem. if there are 1 million results for a query, you'll retrieve them all from the DB to finally return the 30 results asked by the user.

in a general way, pagination is performed in the database, not in memory once you retrieved all the possible results.

of course, rights management makes it a bit more difficult to do. can you think of a way to handle it?

File: entity-service/src/main/kotlin/com/egm/stellio/entity/web/EntityHandler.kt (hunk 1 above)
Comment (HoucemKacem) [reply entity-service/src/main/kotlin/com/egm/stellio/entity/web/EntityHandler.kt]: Limit and page, both, must be strictly greater than 0, no ?

File: entity-service/src/main/kotlin/com/egm/stellio/entity/web/EntityHandler.kt (hunk 2 above)
Comment (HoucemKacem) [reply entity-service/src/main/kotlin/com/egm/stellio/entity/web/EntityHandler.kt]: Done

File: entity-service/src/main/kotlin/com/egm/stellio/entity/web/EntityHandler.kt (hunk 1 above)
Comment (bobeal) [reply entity-service/src/main/kotlin/com/egm/stellio/entity/web/EntityHandler.kt]: yes, strictly greater.

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/web/SubscriptionHandler.kt [hunk 3]
```diff
@@ -72,43 +81,35 @@ class SubscriptionHandler(
...
         val includeSysAttrs = options.filter { it.contains("sysAttrs") }.isPresent
         val contextLink = getContextFromLinkHeaderOrDefault(httpHeaders)
         val mediaType = getApplicableMediaType(httpHeaders)
         if (limit <= 0 || page <= 0)
             return ResponseEntity.status(HttpStatus.BAD_REQUEST).contentType(MediaType.APPLICATION_JSON)
                 .body(BadRequestDataResponse("Page number and Limit must be greater than zero"))
 
+        if (limit > paginationMaxLimit)
+            return ResponseEntity.status(HttpStatus.BAD_REQUEST).contentType(MediaType.APPLICATION_JSON)
+                .body(BadRequestDataResponse("Maximum limit is $paginationMaxLimit"))
```
Comment (bobeal) [inline subscription-service/src/main/kotlin/com/egm/stellio/subscription/web/SubscriptionHandler.kt:98]: you could have something more user friendly ... "You asked for X results, but the supported maximum limit is Y"

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/web/SubscriptionHandler.kt [hunk 4]
```diff
@@ -39,6 +40,14 @@ class SubscriptionHandler(
     private val subscriptionEventService: SubscriptionEventService
 ) {
 
+    companion object {
+        @Value("\${pagination.limit.default:}")
+        private const val paginationDefaultLimit: Int = 30
```
Comment (bobeal) [inline subscription-service/src/main/kotlin/com/egm/stellio/subscription/web/SubscriptionHandler.kt:45]: does the value from the configuration really override the default?

File: subscription-service/src/main/resources/application.properties [hunk 5]
```diff
@@ -31,3 +31,7 @@ application.authentication.enabled = false
 # application.graylog.host = localhost
 # application.graylog.port = 12201
 # application.graylog.source = stellio-int
+
+# Pagination config for query resources endpoints
+pagination.limit.default = 30
```
Comment (bobeal) [inline subscription-service/src/main/resources/application.properties:36]: it is cleaner to have our own keys under the `application` prefix, to better distinguish them from external properties brought by Spring and others

File: subscription-service/src/main/kotlin/com/egm/stellio/subscription/web/SubscriptionHandler.kt (hunk 4 above)
Comment (bobeal) [reply subscription-service/src/main/kotlin/com/egm/stellio/subscription/web/SubscriptionHandler.kt:45]: also, using a configuration properties class is cleaner. see https://github.com/stellio-hub/stellio-context-broker/blob/feature/load-test-experiments/search-service/src/main/kotlin/com/egm/stellio/search/config/ApplicationProperties.kt for instance.

File: entity-service/src/main/kotlin/com/egm/stellio/entity/repository/SearchRepository.kt [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package com.egm.stellio.entity.repository
+
+import java.net.URI
+
+interface SearchRepository {
+
+    fun getEntities(params: Map<String, Any?>, userId: String, page: Int, limit: Int): Pair<Int, List<URI>>
```
Comment (bobeal) [inline entity-service/src/main/kotlin/com/egm/stellio/entity/repository/SearchRepository.kt:7]: what is the Int in the returned type?

File: entity-service/src/main/kotlin/com/egm/stellio/entity/repository/StandaloneNeo4jSearchRepository.kt [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+import java.net.URI
+import java.util.regex.Pattern
+
+@Component
+@ConditionalOnProperty("application.authentication.enabled", havingValue = "false")
+class StandaloneNeo4jSearchRepository(
+    private val session: Session
+) : SearchRepository {
+
+    override fun getEntities(params: Map<String, Any?>, userId: String, page: Int, limit: Int): Pair<Int, List<URI>> {
```
Comment (bobeal) [inline entity-service/src/main/kotlin/com/egm/stellio/entity/repository/StandaloneNeo4jSearchRepository.kt:18]: the two implementations (with ou without authentication enabled) are finally very similar, no?

File: entity-service/src/main/kotlin/com/egm/stellio/entity/repository/SearchRepository.kt (hunk 6 above)
Comment (HoucemKacem) [reply entity-service/src/main/kotlin/com/egm/stellio/entity/repository/SearchRepository.kt:7]: The int is the total number of results.

File: entity-service/src/main/kotlin/com/egm/stellio/entity/repository/StandaloneNeo4jSearchRepository.kt (hunk 7 above)
Comment (HoucemKacem) [reply entity-service/src/main/kotlin/com/egm/stellio/entity/repository/StandaloneNeo4jSearchRepository.kt:18]: Yes, but not for all clauses.

**Review by bobeal — APPROVED** [review]

**Comment by HoucemKacem** [comment]
@bobeal can you check [this](https://github.com/stellio-hub/stellio-context-broker/pull/392/commits/302139a593c2b47e9b8d705bd1c74aed09619e76) commit to return allowed entities as per the specific access policy.

File: entity-service/src/test/kotlin/com/egm/stellio/entity/authorization/Neo4jAuthorizationRepositoryTest.kt [hunk 8]
```diff
@@ -16,13 +17,15 @@ import com.egm.stellio.shared.util.JsonLdUtils.EGM_SPECIFIC_ACCESS_POLICY
...
 import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.boot.context.properties.EnableConfigurationProperties
 import org.springframework.boot.test.context.SpringBootTest
 import org.springframework.context.annotation.Import
 import org.springframework.test.context.ActiveProfiles
 import java.net.URI
 
 @SpringBootTest
 @ActiveProfiles("test")
+@EnableConfigurationProperties(ApplicationProperties::class)
```
Comment (bobeal) [inline entity-service/src/test/kotlin/com/egm/stellio/entity/authorization/Neo4jAuthorizationRepositoryTest.kt:28]: it is already configured in the main class, it should not appear here

File: entity-service/src/test/kotlin/com/egm/stellio/entity/authorization/Neo4jAuthorizationRepositoryTest.kt [hunk 9]
```diff
@@ -45,6 +48,7 @@ class Neo4jAuthorizationRepositoryTest {
     companion object {
         const val EGM_CAN_READ = "https://ontology.eglobalmark.com/authorization#rCanRead"
         const val EGM_CAN_WRITE = "https://ontology.eglobalmark.com/authorization#rCanWrite"
+        const val EGM_CAN_ADMIN = "https://ontology.eglobalmark.com/authorization#rCanAdmin"
```
Comment (bobeal) [inline entity-service/src/test/kotlin/com/egm/stellio/entity/authorization/Neo4jAuthorizationRepositoryTest.kt:51]: unused?

Comment (HoucemKacem) [reply entity-service/src/test/kotlin/com/egm/stellio/entity/authorization/Neo4jAuthorizationRepositoryTest.kt:51]: used in Neo4jSearchRepositoryTests, L83

**Pull request merged**